<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>get.SINBA.DE.2G — get.SINBA.DE.2G • SINBA</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="get.SINBA.DE.2G — get.SINBA.DE.2G"><meta name="description" content="Differential Expression Analysis and Differential Activity Analysis Between 2 Sample Groups. get.SINBA.DE.2G is a function performs differential gene expression analysis and differential driver activity analysis between control group (parameter G0) and experimental group (parameter G1), using different methods."><meta property="og:description" content="Differential Expression Analysis and Differential Activity Analysis Between 2 Sample Groups. get.SINBA.DE.2G is a function performs differential gene expression analysis and differential driver activity analysis between control group (parameter G0) and experimental group (parameter G1), using different methods."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SINBA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorial" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorial</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorial"><li><a class="dropdown-item" href="../tutorials/sinba.html">SINBA Drug database</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../tutorials/TALL.html">Example combination therapy in TALL</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../articles/paper.html">Articles</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/jyyulab/SINBA"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>get.SINBA.DE.2G</h1>

      <div class="d-none name"><code>get.SINBA.DE.2G.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Differential Expression Analysis and Differential Activity Analysis Between 2 Sample Groups. <code>get.SINBA.DE.2G</code> is a function performs differential gene expression analysis and differential driver activity analysis between control group (parameter G0) and experimental group (parameter G1), using different methods.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get.SINBA.DE.2G</span><span class="op">(</span></span>
<span>  eset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  G1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  G0 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  G1_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  G0_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  do.log2 <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  test.use <span class="op">=</span> <span class="st">"limma"</span>,</span>
<span>  random_effect <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-eset">eset<a class="anchor" aria-label="anchor" href="#arg-eset"></a></dt>
<dd><p>ExpressionSet class object, contains gene expression data or driver activity data.</p></dd>


<dt id="arg-g-">G1<a class="anchor" aria-label="anchor" href="#arg-g-"></a></dt>
<dd><p>a vector of characters, the sample names of experimental group.</p></dd>


<dt id="arg-g-">G0<a class="anchor" aria-label="anchor" href="#arg-g-"></a></dt>
<dd><p>a vecotr of characters, the sample names of control group.</p></dd>


<dt id="arg-g-name">G1_name<a class="anchor" aria-label="anchor" href="#arg-g-name"></a></dt>
<dd><p>character, the name of experimental group (e.g. "Male"). Default is "G1".</p></dd>


<dt id="arg-g-name">G0_name<a class="anchor" aria-label="anchor" href="#arg-g-name"></a></dt>
<dd><p>character, the name of control group (e.g. "Female"). Default is "G0".</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical, if TRUE, sample names of both groups will be printed. Default is TRUE.</p></dd>


<dt id="arg-test-use">test.use<a class="anchor" aria-label="anchor" href="#arg-test-use"></a></dt>
<dd><p>character, users can choose one of the methods c("limma","t.test","wilcox"). Default is "limma"</p></dd>


<dt id="arg-random-effect">random_effect<a class="anchor" aria-label="anchor" href="#arg-random-effect"></a></dt>
<dd><p>a vector of characters, vector or factor specifying a blocking variable.Default is NULL, no random effect will be considered.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Return a data frame. Rows are genes/drivers, columns are "ID","Z-statistics","P.Value","adj.P.Val","logFC"</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">phe_info</span> <span class="op">&lt;-</span> <span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">SINBA.par</span><span class="op">$</span><span class="va">cal.eset</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'object' in selecting a method for function 'pData': object 'SINBA.par' not found</span>
<span class="r-in"><span><span class="va">phe_info</span><span class="op">$</span><span class="va">subgroup</span><span class="op">&lt;-</span><span class="va">phe_info</span><span class="op">$</span><span class="va">das_sensitivity_category1</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'phe_info' not found</span>
<span class="r-in"><span><span class="va">G0</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">phe_info</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">phe_info</span><span class="op">$</span><span class="va">`subgroup`</span><span class="op">==</span><span class="st">"resistant"</span><span class="op">)</span><span class="op">]</span> <span class="co"># get sample list for G0</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'phe_info' not found</span>
<span class="r-in"><span><span class="va">G1</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">phe_info</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">phe_info</span><span class="op">$</span><span class="va">`subgroup`</span><span class="op">==</span><span class="st">"sensitive"</span><span class="op">)</span><span class="op">]</span> <span class="co"># get sample list for G1</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'phe_info' not found</span>
<span class="r-in"><span><span class="va">ac_eset</span><span class="op">&lt;-</span><span class="fu"><a href="generate.SINBA.eset.html">generate.SINBA.eset</a></span><span class="op">(</span>exp_mat<span class="op">=</span><span class="va">ac_mat</span>, phenotype_info<span class="op">=</span><span class="va">phe_info</span>, feature_info<span class="op">=</span><span class="cn">NULL</span>, annotation_info<span class="op">=</span><span class="st">""</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'ac_mat' not found</span>
<span class="r-in"><span><span class="va">DE_gene_limma</span> <span class="op">&lt;-</span> <span class="fu">get.SINBA.DE.2G</span><span class="op">(</span>eset<span class="op">=</span><span class="va">SINBA.par</span><span class="op">$</span><span class="va">cal.eset</span>,G1<span class="op">=</span><span class="va">G1</span>,G0<span class="op">=</span><span class="va">G0</span>,G1_name<span class="op">=</span><span class="st">"sensitive"</span>,G0_name<span class="op">=</span><span class="st">'resistant'</span>,test.use<span class="op">=</span><span class="st">"limma"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'SINBA.par' not found</span>
<span class="r-in"><span><span class="va">DA_gene_limma</span> <span class="op">&lt;-</span> <span class="fu">get.SINBA.DE.2G</span><span class="op">(</span>eset<span class="op">=</span><span class="va">ac_eset</span>,G1<span class="op">=</span><span class="va">G1</span>,G0<span class="op">=</span><span class="va">G0</span>,G1_name<span class="op">=</span><span class="st">"sensitive"</span>,G0_name<span class="op">=</span><span class="st">'resistant'</span>,test.use<span class="op">=</span><span class="st">"limma"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'ac_eset' not found</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jingjing Liu, Yu Lab @ St Jude Children’s Research Hospital.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

